Statically compile libgit2 into KraftKit

nderjung opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [x] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Statically compile libgit2 into kraftkit.  This has been a major source of frustration for newcomers to the project who must use a specific libgit2 shared object dependency in auxiliary to the supplied `kraft` binary.

To mitigate against this, statically compile libgit2 with KraftKit.

GitHub-Fixes: #179
GitHub-Closes: #31

nderjung committed:
Makefile: Rename `deps` target to `tidy`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
Statically link libgit2-dev

This commit statically links binaries of KraftKit with libgit2.  To do
this we vendor the git2go package which requires libgit2.  However, to
prevent vendoring every required Go module, we a). include git2go as a
Git submodule and b). explicitly ignore the vendor directory in the `go
build` command by passing `-mod=readonly` flag.

This commit simultaneously ignores the vendor directory in the CI
configuration and makes cloning the libgit2 repository a dependency for
when starting the developer environment which would otherwise result in
a dubious ownership of the submodule if cloned with a different user and
group ID.

GitHub-Fixes: #179
GitHub-Closes: #31
Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
Squash the license and copyright notice preamble

This commit "squashes" the license and copyright preamble in files
associated with this sign off.  The preamble already includes the SPDX
license identifier which essentially duplicates the text in question.
Additionally, this license, in accordance with the license, is
distributed with the source code via the `LICENSE.md` file which is
accompanied in this repository.

Additionally, copyright notice is provided to The KraftKit Authors,
which entitles new developers who contribute commits to the project an
associated copyright with regard to their contributions without having
to specifically update the license and copyright preamble.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
.vscode, .gitignore: Include VS Code settings

This commit includes workspace settings for VS Code, particularly
configuration which manipulates the VS Code Extension for Go.  Due to
partial vendoring of the project, we must signify to Go to use
`-mod=readonly` in both the build process (if executed through VS
Code itself) and to the Go language service which would otherwise error
and prevent any in-IDE utilities such as code-completion.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on .github/workflows/gochecks.yaml:
```diff
@@ -22,8 +22,12 @@ jobs:
+        run: |
+          apt-get update;
+          apt-get install -y --no-install-recommends \
+            cmake \
+            libgit2-dev;
```
```suggestion
            cmake;
```

No longer needed.

craciunoiuc commented on .github/workflows/release-stable.yaml:
```diff
@@ -32,9 +32,12 @@ jobs:
+        run: |
+          apt-get update;
+          apt-get install -y --no-install-recommends \
+            cmake \
+            libgit2-dev;
```
```suggestion
            cmake;
```

No longer needed.

craciunoiuc commented on .github/workflows/release-staging.yaml:
```diff
@@ -32,9 +32,12 @@ jobs:
+        run: |
+          apt-get update;
+          apt-get install -y --no-install-recommends \
+            cmake \
+            libgit2-dev;
```
```suggestion
            cmake;
```

No longer needed.

craciunoiuc commented on Makefile:
```diff
@@ -162,3 +141,15 @@ properclean: IMAGE       ?= $(REGISTRY)/$(ORG)/$(REPO)/$(ENVIRONMENT):$(IMAGE_TA
+	$(GIT) -C $(VENDORDIR)/github.com/libgit2/git2go/v31 submodule update --init --recursive
+
+$(VENDORDIR)/github.com/libgit2/git2go:
+	$(GIT) clone --branch v31.7.9 --recurse-submodules https://github.com/libgit2/git2go.git $@/v31
```
```suggestion
	$(GIT) clone --branch $(GIT2GOBRANCH) --recurse-submodules https://github.com/libgit2/git2go.git $@/v31
```

We should probably make it a variable so it is easier to change/we don't forget about it.

nderjung replied:
Then we need a bit of Makefile magic to strip the `.7.9` out of the `v31.7.9` variable value, probably it is fine to leave as is for now

craciunoiuc commented on buildenvs/myself.Dockerfile:
```diff
@@ -42,6 +42,7 @@ RUN set -xe; \
     apt-get install -y --no-install-recommends \
       build-essential \
       libgit2-dev \
+      cmake \
```
```suggestion
      cmake \
```

No longer needed.

craciunoiuc commented on .gitignore:
```diff
@@ -1,9 +1,6 @@
 # Ignore distributables folder
 dist/
 
-# Ignore vscode folder
-.vscode/
-
```
No need to remove the ignore to the vscode server. We can ignore it and add the changes at the same time.
For example I have some logs in the folder that are autogenerated and they pollute the `git status` command.

![image](https://user-images.githubusercontent.com/46280801/218682605-cd56812f-ab7a-4620-8eed-855d7ff24908.png)

If we don't ignore it, there might also be a change to add these files accidentally.
I say we ignore it, but keep the `settings.json` change.

craciunoiuc reviewed (changes requested):
An initial set of comments from my side

nderjung committed:
Initialize go workspace

This commit introduces a new top-level file `go.work`, an axuiliary file
made available from Go 1.18 which signifies a repository with multiple
Go modules.  Since there are essentially 3 modules within KraftKit (so
far) we introduce this file to aid the Go compiler and Go tools
navigating the repository.

In addition to this, given that we are now partially vendoring the
repository due to the inclusion of libgit2, this file helps in the
vendoring-realization process for Go tools such as with `gopls`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
buildenvs: Mark the Git repository as safe

To prevent issues with regard to dubious ownership through partial
vendoring via git submodules, mark the Git repository as safe in the
build environment.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc reviewed (approved):
All good here. Thanks!

Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
